aptos-move/aptos-aggregator/src/delta_change_set.rs

```diff
@@ -76,27 +76,66 @@ impl DeltaOp {
         }
     }
 
-    /// Aggregates another delta into `self`.
-    pub fn merge_with(&mut self, other: DeltaOp) -> PartialVMResult<()> {
+    /// Applies self on top of previous delta, merging them together. Note
+    /// that the strict ordering here is crucial for catching overflows
+    /// correctly.
+    pub fn merge_onto(&mut self, previous_delta: DeltaOp) -> PartialVMResult<()> {
         use DeltaUpdate::*;
 
-        Ok(match (self.update, other.update) {
-            (Plus(value), Plus(other_value)) => {
-                let new_value = addition(value, other_value, self.limit)?;
-                self.update = Plus(new_value);
-            }
-            (Plus(value), Minus(other_value)) | (Minus(other_value), Plus(value)) => {
-                if value >= other_value {
-                    let new_value = subtraction(value, other_value)?;
-                    self.update = Plus(new_value);
-                } else {
-                    let new_value = subtraction(other_value, value)?;
-                    self.update = Minus(new_value);
+        Ok(match previous_delta.update {
+            Plus(value) => {
+                // Before we proceed to merging deltas, make sure we verify that
+                // `self` delta did not overflow when shifted by +value. At the
+                // same time, we can compute new history values for `self` delta.
+                let other_max_positive = addition(value, self.max_positive, self.limit)?;
```

Comment:
Why do we default to a value in the next statement but don't default here? What happens if this overflows, especially for critical use such as for total supply?

Reply:
In these 2 statements:
```rust
let other_max_positive = addition(value, self.max_positive, self.limit)?;          // (1)
let other_min_negative = subtraction(self.min_negative, value).or(Ok(0))?;   // (2)
```
we check if history is valid if shifted by `+value`. This is true if `value + max_positive` does not overflow (1) (note that we use `?` to return an error immediately). However, (2) should always succeed, right? `min_negative` is the smallest value delta had seen (e.g. if it was -100 at some point this would be 100), and clearly if shifted up by value we is still correct (e.g. if -100 is ok, clearly something like +30-100 is also ok).

Why default? Well, there can be a case when the value by which we shift is greater than the minimum negative value. Coming back to our example with -100, it is the same as if the value was +101: the min negative would be +101-100=+1 which does not make sense so should be capped at 0.

Same applied for validating `Minus`.

Does this makes sense? Maybe I should extend the comment to mention that cap explicitly...